frontend/lib/hooks/useStreamMessage.tsx

```diff
@@ -106,12 +121,145 @@ function useStreamMessage(cleanlabEnabled: boolean = true) {
   const setThreadStatus = useMessagesStore(state => state.setThreadStatus)
 
   const addHistoryThread = useRagAppStore(state => state.addHistoryThread)
+  const updateHistoryThread = useRagAppStore(state => state.updateHistoryThread)
   const [appSettings] = useAppSettings()
 
   // Always use latest cleanlabEnabled in callbacks
   const cleanlabRef = useRef(cleanlabEnabled)
   cleanlabRef.current = cleanlabEnabled
 
+  const getOrInitBuffer = useCallback((threadId: string) => {
+    let buffer = threadBuffers.get(threadId)
+    if (!buffer) {
+      buffer = []
+      threadBuffers.set(threadId, buffer)
+    }
+    return buffer
+  }, [])
+
```

Comment:
A lot of the functions in here could be defined outside of the hook to avoid needing `useCallback`